spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/BlockingQueueConsumer.java

```diff
@@ -749,22 +747,30 @@ public void handleCancelOk(String consumerTag) {
 			if (logger.isDebugEnabled()) {
 				logger.debug("Received cancellation notice for tag " + consumerTag + "; " + BlockingQueueConsumer.this);
 			}
-			synchronized (BlockingQueueConsumer.this.consumerTags) {
-				BlockingQueueConsumer.this.consumerTags.remove(consumerTag);
-			}
+			BlockingQueueConsumer.this.cancelReceived.set(true);
+			BlockingQueueConsumer.this.consumerTags.remove(consumerTag);
 		}
 
 		@Override
 		public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body)
 				throws IOException {
-			if (logger.isDebugEnabled()) {
-				logger.debug("Storing delivery for " + BlockingQueueConsumer.this);
-			}
-			try {
-				BlockingQueueConsumer.this.queue.put(new Delivery(consumerTag, envelope, properties, body));
+			long deliveryTag = envelope.getDeliveryTag();
+			if (cancelled()) {
```

Comment:
I don't believe this is necessary - we'll never get a delivery after a `cancel` or `cancelOk`, but I suppose it doesn't hurt, except for the (small) overhead.